### DISCUSSION TITLE
How to publish event from webapi server to elsa server

### DISCUSSION THREAD

**Discussion by yopichy** [elsa-workflows/elsa-core#2606] [Q&A]
I have registered azure bus `AddElsa(elsa => elsa.UseAzureServiceBus("connectionstring")` in Elsa Server and already generated queues and topics on azure.
I'm followign tutorial from Medium Article Workflow Driven with Elsa 2, 
![image](https://user-images.githubusercontent.com/2038089/145035508-24b49a1c-c706-49cd-8a43-3de1bb1fc57c.png)
there is a section where the service will publish events using MediatR, can it be replaced with something else, for example MassTransit? If possible, how is the configuration (queue and topic using by elsa) in WebAPI  Server to publish events so that it can be handled by Elsa Server consume Azure Bus?
![image](https://user-images.githubusercontent.com/2038089/145035757-af545eca-ee73-4575-8736-51b1efac028c.png)

**Comment by sfmskywalker** [comment]
MediatR is used to publish domain events, so I wouldn't replace that with an "out of process" mechanism such as MassTransit. What you can do, however, is implement a MediatR notification handler and publish messages via e.g. MassTransit from there. In fact, this is what the core library does as well for some domain events that need to be distributed across nodes in a cluster.

Example from: https://github.com/elsa-workflows/elsa-core/blob/master/src/core/Elsa.Core/Handlers/PublishWorkflowDefinitionEvents.cs

```csharp
public class PublishWorkflowDefinitionEvents : INotificationHandler<WorkflowDefinitionPublished>, INotificationHandler<WorkflowDefinitionRetracted>, INotificationHandler<WorkflowDefinitionDeleted>
{
    private readonly IEventPublisher _eventPublisher;
    public PublishWorkflowDefinitionEvents(IEventPublisher eventPublisher) => _eventPublisher = eventPublisher;
    public Task Handle(WorkflowDefinitionPublished notification, CancellationToken cancellationToken) => _eventPublisher.PublishAsync(notification);
    public Task Handle(WorkflowDefinitionRetracted notification, CancellationToken cancellationToken) => _eventPublisher.PublishAsync(notification);
    public Task Handle(WorkflowDefinitionDeleted notification, CancellationToken cancellationToken) => _eventPublisher.PublishAsync(notification);
}
```

As you can see, `PublishWorkflowDefinitionEvents ` is a notification handler of various domain events, basically forwarding each event to the service bus via `IEventPublisher` which is an abstraction around Rebus, which can be configured to use any service bus broker you want, including Azure Service Bus.

Speaking of Azure Service Bus, you could implement a notification handler like above, but instead of using `IEventPublisher`, you could inject `IQueueMessageSenderFactory` to obtain a `ISenderClient` and send a message using that.

It would be up to you to then also implement Azure Service Bus message workers. The `Elsa.Activities.AzureServiceBus` project has the necessary infrastructure, but I noticed it does not properly expose worker primitives to allow external code to implement workers easily. But we can change that if you think it would be useful. 

Alternatively, you could take advantage of the Rebus abstractions to send & handle messages to the workflow server.
